pkg/addon/addon_update.go

```diff
@@ -0,0 +1,320 @@
+	}
+	return updating, nil
+}
+
+func (c *AddonUpdate) addAddonToCache(addon *addonv1.Addon) {
+	var version = Version{
+		Name:        addon.GetName(),
+		Namespace:   addon.GetNamespace(),
+		PackageSpec: addon.GetPackageSpec(),
+		PkgPhase:    addon.GetInstallStatus(),
+	}
+	c.versionCache.AddVersion(version)
+	c.log.WithValues("addAddonToCache", fmt.Sprintf("Adding %s/%s package %s version cache phase %s", addon.GetNamespace(), addon.GetName(), addon.GetPackageSpec(), version.PkgPhase))
+}
+
+// UpdateAddonStatusLifecycle
+func (c *AddonUpdate) UpdateAddonStatusLifecycle(ctx context.Context, namespace, name string, lifecycle string, lifecyclestatus wfv1.WorkflowPhase) error {
+
+	key := fmt.Sprintf("%s/%s", namespace, name)
+	latest, err := c.getExistingAddon(ctx, key)
+	if err != nil || latest == nil {
+		return err
+	}
+	updating := latest.DeepCopy()
+	prevStatus := latest.Status
+
+	// addon being deletion, skip non-delete wf update
+	if lifecycle != "delete" &&
+		prevStatus.Lifecycle.Installed == addonv1.Deleting {
+		return nil
+	}
+
+	newStatus := addonv1.AddonStatus{
+		Lifecycle: addonv1.AddonStatusLifecycle{},
+		Resources: []addonv1.ObjectStatus{},
+	}
+	newStatus.Reason = prevStatus.Reason
+	newStatus.Resources = append(newStatus.Resources, prevStatus.Resources...)
+	newStatus.Checksum = prevStatus.Checksum
+	newStatus.StartTime = prevStatus.StartTime
+	if lifecycle == "prereqs" {
+		newStatus.Lifecycle.Prereqs = addonv1.ApplicationAssemblyPhase(lifecyclestatus)
+		newStatus.Lifecycle.Installed = prevStatus.Lifecycle.Installed
+		if newStatus.Lifecycle.Prereqs == addonv1.Failed {
+			newStatus.Lifecycle.Installed = addonv1.Failed
+			newStatus.Reason = "prereqs wf fails."
+		}
+	} else if lifecycle == "install" || lifecycle == "delete" {
+		newStatus.Lifecycle.Installed = addonv1.ApplicationAssemblyPhase(lifecyclestatus)
+		newStatus.Lifecycle.Prereqs = prevStatus.Lifecycle.Prereqs
+		if addonv1.ApplicationAssemblyPhase(lifecyclestatus) == addonv1.Succeeded {
+			newStatus.Reason = ""
+		}
+
+		// check whether need patch complete
+		if lifecycle == "install" && newStatus.Lifecycle.Installed.Completed() {
+			labels := updating.GetLabels()
+			if labels == nil {
+				labels = map[string]string{}
+			}
+
+			updating.Status = newStatus
+			if _, err := c.updateAddon(ctx, updating); err != nil {
+				return err
+			}
+			return nil
+		}
+	}
+	updating.Status = newStatus
+
+	if lifecycle == "delete" && addonv1.ApplicationAssemblyPhase(lifecyclestatus).Succeeded() {
+		if prevStatus.Lifecycle.Installed.Completed() || prevStatus.Lifecycle.Installed.Deleting() {
+			c.removeFinalizer(updating)
+			if _, err := c.updateAddon(ctx, updating); err != nil {
+				return err
+			}
+			c.removeFromCache(updating.Name)
+			return nil
+		}
+	}
+
+	var afterupdating *addonv1.Addon
+	patchLabel := false
+	if lifecycle == "install" && addonv1.ApplicationAssemblyPhase(lifecyclestatus).Succeeded() {
```

Comment:
Use lifecyclestep constant

Reply:
73e945ec51f9bb0089e30d4761e301c3d3cd52a0

Reply:
```suggestion
	if lifecycle == v1alpha1.Install  && addonv1.ApplicationAssemblyPhase(lifecyclestatus).Succeeded() {
```

Use constant https://github.com/keikoproj/addon-manager/blob/5830407fb2242fd09621c8f9ba86f50ad18d6b29/api/addon/v1alpha1/addon_types.go#L171-L180

Reply:
07903cafbefbde5e80a681630e88d2e6f75aace1